river/drift/adwin_c.pyx

```diff
@@ -0,0 +1,379 @@
+            If True then a change is detected.
+
+        """
+        return self._update(value)
+
+    cdef bint _update(self, double value):
+
+        cdef double incremental_variance
+
+        self.width += 1
+
+        self._insert_element(value, 0.0)
+        incremental_variance = 0.0
+        if self.width > 1.0:
+            incremental_variance = (
+                    (self.width - 1.0)
+                    * (value - self.total / (self.width - 1.0))
+                    * (value - self.total / (self.width - 1.0))
+                    / self.width
+            )
+        self.variance += incremental_variance
+        self.total += value
+        self._compress_buckets()
+
+        return self._detect_change()
+
+    cdef void _insert_element(self, double value, double variance):
+        cdef Bucket bucket = self.bucket_deque[0]
+        bucket.insert_data(value, variance)
+        self.n_buckets += 1
+
+        if self.n_buckets > self.max_n_buckets:
+            self.max_n_buckets = self.n_buckets
+
+    cdef double _bucket_size(self, int row):
+        return pow(2, row)
+
+    cdef double _delete_element(self):
+        """Delete an item from the bucket deque.
+
+        Deletes the last item and updates relevant statistics kept by ADWIN.
+
+        Returns
+        -------
+        The size of the item being dropped
+
+        """
+        cdef:
+            double n1, u1
+            Bucket bucket
+            double incremental_variance
+
+        bucket = self.bucket_deque[-1]
+        n1 = self._bucket_size(len(self.bucket_deque) - 1)
+        self.width -= n1
+        self.total -= bucket.get_total_at(0)
+        u1 = bucket.get_total_at(0) / n1
+        incremental_variance = (
+                bucket.get_variance_at(0) + n1 * self.width
+                * (u1 - self.total / self.width) * (u1 - self.total / self.width)
+                / (n1 + self.width)
+        )
+        self.variance -= incremental_variance
+        bucket.remove()
+        self.n_buckets -= 1
+
+        if bucket.current_idx == 0:
+            self.bucket_deque.pop()
+
+        return n1
+
+    cdef void _compress_buckets(self):
+
+        cdef:
+            int idx, k
+            double n1, n2, u1, u2, incremental_variance
+            Bucket bucket, next_bucket
+
+        for idx, bucket in enumerate(self.bucket_deque):
+            k = bucket.current_idx
+            # If the row is full, merge buckets
+            if k == self.MAX_BUCKETS + 1:
+                try:
+                    next_bucket = self.bucket_deque[idx + 1]
+                except IndexError:
+                    self.bucket_deque.append(Bucket())
+                    next_bucket = self.bucket_deque[-1]
+                n1 = self._bucket_size(idx)
+                n2 = self._bucket_size(idx)
+                u1 = bucket.get_total_at(0) / n1
+                u2 = bucket.get_total_at(1) / n2
+                incremental_variance = n1 * n2 * ((u1 - u2) * (u1 - u2)) / (n1 + n2)
```

Comment:
Here too. Apparently, an approximation is being used. Note that the formula only manipulates the mean estimators, not the actual variance.

Reply:
Addressed.